Implement PRTB, CRTB and RoleTemplate handlers using Cluster Role Aggregation

JonCrowther opened the pull request:
## Issue: https://github.com/rancher/rancher/issues/46077
 
## Problem
Rancher RBAC resources make up the largest number of resources in the local and downstream clusters, especially if there's a lot of role inheritance. That can become a problem in larger systems as it will eventually overwhelm etcd.

## Solution
Using [cluster role aggregation](https://kubernetes.io/docs/reference/access-authn-authz/rbac/#aggregated-clusterroles) we can reduce the overall number of resources and simplify the existing handlers.

Overall, Role Templates will now create Cluster Roles, and then PRTBs/CRTBs will create Cluster Role Bindings and Role Bindings. Previously, the CRTB/PRTB handler was responsible for creating both Roles and Role Bindings.

All changes are gated behind the feature flag `aggregated-roletemplates`. For this to work, Rancher must be installed from scratch with the feature flag set to `true`. It currently doesn't support being set after Rancher is already running. Migration is future work.

It is a very large PR, so I tried to split it in the commits. The breakdown is:
1. Add feature flag. Create project and cluster membership roles as soon as the cluster/project is created instead of waiting for a PRTB or CRTB to be created.
2. Add common functions that are used by the handlers
3. Role Template handlers + tests
4. PRTB handlers + tests
5. CRTB handlers + tests
6. Register the new handlers
 
## Testing
Added lots of unit tests for PRTBs, CRTBs, RoleTemplates, and common functions.

## Engineering Testing
### Manual Testing
<!-- Describe what manual testing you did (if no testing was done, explain why). -->

### Automated Testing
<!-- Ensure there are unit/integration/validation tests added (if possible); describe what cases they cover and do not cover. -->
* Test types added/modified:
    * Unit
    * Integration (Go Framework)
    * Integration (v2prov Framework)
    * Validation (Go Framework)
    * Other - Explain: _EXPLAIN_
    * None
    * _REMOVE NOT APPLICABLE BULLET POINTS ABOVE_
* If "None" - Reason: _EXPLAIN THE REASON_
  <!-- 
  Non-exhaustive list of reasons:
    - Lack of the framework capable of testing this fix/change
    - Tight deadlines / critical priority to get fix/change in - !ensure GH issue is logged to add tests!
    - No application logic is modified by this change, e.g. refactoring/cosmetic/non-code/test change
    - Tests implemented in another PR elsewhere - !ensure GH PR link is added!
    - Other (explain)
  Note: Outside of the exceptions above, the "existing tests cover the changes" is very unlikely to be an acceptable reason as the existing tests generally don't cover the logic changes implemented by this PR 
  -->
* If "None" - GH Issue/PR: _LINK TO GH ISSUE/PR TO ADD TESTS_

Summary: _TODO_

## QA Testing Considerations
<!-- Highlight areas or (additional) cases that QA should test w.r.t a fresh install as well as the upgrade scenarios -->
 
### Regressions Considerations
<!-- Dedicated section to specifically call out any areas that with higher chance of regressions caused by this change, include estimation of probability of regressions -->
_TODO_

Existing / newly added automated tests that provide evidence there are no regressions:
* _TODO_

aruiz14 commented on pkg/controllers/managementuser/rbac/roletemplates/common.go:
```diff
@@ -0,0 +1,75 @@
+
+// deleteServiceAccountImpersonator checks if there are any CRBTs or PRTBs for this user. If there are none, remove their Service Account Impersonator.
+func (ih *impersonationHandler) deleteServiceAccountImpersonator(username string) error {
+	lo := metav1.ListOptions{FieldSelector: "userName=" + username}
```
This fails due to Field Selectors only available for a [small set of fields](https://kubernetes.io/docs/concepts/overview/working-with-objects/field-selectors/#supported-fields) (although this will change in Kubernetes >= 1.31).
We'll need to use a similar approach to what [other implementations do](https://github.com/rancher/rancher/blob/de9dda187837850ba4214cb4e37069d987bdc22b/pkg/controllers/managementuser/rbac/impersonation_handler.go), and use a custom indexer, for example. In fact, can we reuse something that exists already?

JonCrowther replied:
I've changed this to use a custom indexer along with a comment recommending custom field selectors once it's fully supported.

JonCrowther commented on pkg/user/zz_manager_fake.go:
```diff
@@ -0,0 +1,206 @@
+//
+// Generated by this command:
+//
+//	mockgen -source=manager.go -destination=zz_manager_fake.go -package=user
```
I can, but I didn't see any previous examples of that in rancher/rancher. I wasn't sure if there was a reason for that so I opted to follow the convention.

raulcabello commented on pkg/controllers/management/auth/roletemplates/crtb_handler.go:
```diff
@@ -0,0 +1,270 @@
+}
+
+// removeClusterRoleBindings removes all cluster role bindings owned by the CRTB
+func (c *crtbHandler) removeClusterRoleBindings(crtb *v3.ClusterRoleTemplateBinding) error {
```
can we use k8s OwnerReferences instead?

JonCrowther replied:
We definitely should be for anything stored in the local cluster.

raulcabello replied:
that makes sense. Can we do it for RoleTemplates that are non-namespaced?

JonCrowther replied:
Roletemplate use OwnerReferences in 1f5e1b4fd6c12a68a9bd068db8a79de05bf4ea65

aruiz14 commented on pkg/controllers/managementuser/rbac/roletemplates/common.go:
```diff
@@ -0,0 +1,80 @@
+
+type impersonationHandler struct {
+	userContext *config.UserContext
+	crtbClient  mgmtv3.ClusterRoleTemplateBindingController
+	crtbCache   mgmtv3.ClusterRoleTemplateBindingCache
+	prtbClient  mgmtv3.ProjectRoleTemplateBindingController
```
`crtbClient` and `prtbClient` clients are no longer needed now, right?

JonCrowther replied:
Addressed in b6c18a7a080b2d7bfef6fb70efa0adaa74034a2c

aruiz14 commented on pkg/controllers/managementuser/rbac/roletemplates/register.go:
```diff
@@ -0,0 +1,56 @@
+const (
+	crtbChangeHandler   = "cluster-crtb-change-handler"
+	crtbRemoveHandler   = "cluster-crtb-remove-handler"
+	crtbUsernameIndexer = "cluster-crtb-username-indexer"
```
It seems that the convention for an "Index" (not Indexer) is to include a "namespace" or suffix (in order to avoid collisions). Example: https://github.com/rancher/rancher/blob/main/pkg/controllers/management/auth/crtb_handler.go#L38

aruiz14 replied:
Indeed, it now panics due to a repeated index name:
```
panic: indexer conflict: map[cluster-crtb-username-indexer:{}]
```

aruiz14 replied:
Ah! The problem is that this runs for every downstream cluster (as part of `managementuser` controllers), but registers the indexers in the upstream cluster caches (so called ScaledContext). Please my comment below about moving this to a `RegisterIndexers` separate function

JonCrowther replied:
Fixed and moved to a `RegisterIndexers` function in 8d6b143e753a900afbc5751c1d1dee668fbe9f8b

aruiz14 commented on pkg/controllers/managementuser/rbac/roletemplates/register.go:
```diff
@@ -0,0 +1,56 @@
+	c := newCRTBHandler(workload)
+	management.Wrangler.Mgmt.ClusterRoleTemplateBinding().OnChange(ctx, crtbChangeHandler, c.OnChange)
+	management.Wrangler.Mgmt.ClusterRoleTemplateBinding().OnRemove(ctx, crtbRemoveHandler, c.OnRemove)
+	management.Wrangler.Mgmt.ClusterRoleTemplateBinding().Cache().AddIndexer(crtbUsernameIndexer, getCRTBByUsername)
```
It seems there are not many wrangler indexers (yet), but there seems to be a convention to add indexers in a `RegisterIndexers` separate function:
https://github.com/rancher/rancher/blob/d58c180728037b5d3090e2262fce3d33cabaf0fa/pkg/controllers/management/auth/register.go#L17

JonCrowther replied:
Done in 8d6b143e753a900afbc5751c1d1dee668fbe9f8b

aruiz14 commented on pkg/controllers/managementuser/rbac/roletemplates/register.go:
```diff
@@ -0,0 +1,56 @@
+	if obj.UserName != "" && obj.ClusterName != "" {
+		return []string{name.SafeConcatName(obj.ClusterName, obj.UserName)}, nil
+	}
+	return []string{}, nil
```
Just out of curiosity, cannot we just return `nil, nil`?

JonCrowther replied:
I'm actually not sure, I just based it on other examples. I'll test and find out.

raulcabello commented on pkg/rbac/common.go:
```diff
@@ -279,3 +300,202 @@ func isRuleInTargetAPIGroup(rule rbacv1.PolicyRule) bool {
+	resource, err := client.Get(obj.GetName(), metav1.GetOptions{})
+	if err != nil {
+		if !apierrors.IsNotFound(err) {
+			return nil
```
should we return the error here? Imagine there is an error getting the resource (e.g. cluster unavailable) if we return nil we would not retry.

JonCrowther replied:
Addressed in b6c18a7a080b2d7bfef6fb70efa0adaa74034a2c

raulcabello commented on pkg/rbac/common.go:
```diff
@@ -279,3 +300,202 @@ func isRuleInTargetAPIGroup(rule rbacv1.PolicyRule) bool {
+			same = false
+			currentCR.AggregationRule = nil
+		}
+		if !reflect.DeepEqual(currentCR.Rules, wantedCR.Rules) {
```
`reflect.DeepEquals() `considers that a nil list is not equal to a empty list. We should always use semantic DeepEqual to compare rules. This has caused bugs in the past (see https://github.com/rancher/rancher/commit/290254928fa1782d482245a640a25c520bcb9616)

JonCrowther replied:
Addressed in b6c18a7a080b2d7bfef6fb70efa0adaa74034a2c

raulcabello commented on pkg/controllers/management/auth/roletemplates/crtb_handler.go:
```diff
@@ -0,0 +1,278 @@
+	return desiredCRBs, nil
+}
+
+// reconcileBindings takes a map of desired ClusterRoleBindings and a slice of already existing ClusterRoleBindings and ensures that all the desired CRBs exist.
```
nit: this comment indicates the desired ClusterRoleBindings and a slice of already existing ClusterRoleBindings are provided in the parameters, but the are retreived inside the method.

JonCrowther replied:
Addressed in b6c18a7a080b2d7bfef6fb70efa0adaa74034a2c

raulcabello commented on pkg/controllers/managementuser/rbac/roletemplates/crtb_handler.go:
```diff
@@ -0,0 +1,178 @@
+
+	// Find if there is a CRB that already exists and delete all excess CRBs
+	var matchingCRB *rbacv1.ClusterRoleBinding
+	for _, currentCRB := range currentCRBs.Items {
```
There should always be only one CRB that matches `LabelSelector: rbac.GetCRTBOwnerLabel(crtb.Name)`, right? 
It'd be great to add a check or a comment, since this for suggest there might be more

JonCrowther replied:
Addressed in b6c18a7a080b2d7bfef6fb70efa0adaa74034a2c

raulcabello commented on pkg/controllers/managementuser/rbac/roletemplates/prtb_handler.go:
```diff
@@ -0,0 +1,245 @@
+		if err != nil {
+			return nil, err
+		}
+		for _, crb := range rbs.Items {
```
nit: it should be rb, not crb

JonCrowther replied:
Addressed in b6c18a7a080b2d7bfef6fb70efa0adaa74034a2c

raulcabello commented on pkg/controllers/managementuser/rbac/roletemplates/roletemplate_handler.go:
```diff
@@ -0,0 +1,204 @@
+
+// addLabelToExternalRole ensures the external role has the right aggregation label.
+// It is a no-op if the RoleTemplate does not have an external role.
+func (rth *roleTemplateHandler) addLabelToExternalRole(rt *v3.RoleTemplate) error {
```
We should think about modifying external `ClusterRoles` that we don't own as discussed yesterday. This can lead to problems with gitops or `ClusterRoles` managed by a another controller.
What if we don't use aggregation for external `RoleTemplates`? I don't think users would add many inherited RoleTemplates to an external RoleTemplates.

JonCrowther replied:
I think that makes sense. I also made the decision that external roles can't provide access to management plane resources. To me it makes sense that external RoleTemplates don't get the added Rancher RBAC functionality.

alegrey91 commented on pkg/controllers/management/auth/project_cluster/common.go:
```diff
@@ -75,3 +80,89 @@ func reconcileResourceToNamespace(obj runtime.Object, controller string, nsListe
+func newOwnerReference(obj runtime.Object) (metav1.OwnerReference, error) {
+	metadata, err := meta.Accessor(obj)
+	if err != nil {
+		return metav1.OwnerReference{}, fmt.Errorf("error accessing object type %v: %w", obj, err)
```
Is it intentional to keep the same message as the line 160?
Wouldn't be better to differentiate them for better debugging?
(Just asking)

JonCrowther replied:
Addressed in b6c18a7a080b2d7bfef6fb70efa0adaa74034a2c

alegrey91 commented on pkg/controllers/management/auth/register.go:
```diff
@@ -86,6 +83,15 @@ func RegisterEarly(ctx context.Context, management *config.ManagementContext, cl
 	management.Management.GlobalRoleBindings("").AddHandler(ctx, "legacy-grb-cleaner", grbLegacy.sync)
 	management.Management.RoleTemplates("").AddHandler(ctx, "legacy-rt-cleaner", rtLegacy.sync)
 	globalroles.Register(ctx, management, clusterManager)
+
```
WDYT to add a small comment to clarify why we are keeping both new and old way?

JonCrowther replied:
Addressed in b6c18a7a080b2d7bfef6fb70efa0adaa74034a2c

alegrey91 commented on pkg/controllers/management/auth/roletemplates/common.go:
```diff
@@ -0,0 +1,158 @@
+	return existingCRB, nil
+}
+
+// deleteMembershipBinding checks if the user is still a member of the Project or Cluster specified by PRTB/CRTB. If they are no longer a member, delete the bindings.
```
move "If they are no longer a member" to "If the user is no longer a member" ?

JonCrowther replied:
Addressed in b6c18a7a080b2d7bfef6fb70efa0adaa74034a2c

JonCrowther commented:
My original change was creating RoleBindings for promoted roles in PRTBs. The promoted roles actually represent non-namespaced resources that PRTBs can give access to, so we should be creating a ClusterRoleBinding instead. I've added the changes in f2389f9a7e07ecf1d3f2faf83a77722d2a9567a3 along with unit tests.

raulcabello commented on pkg/controllers/management/auth/project_cluster/common.go:
```diff
@@ -75,3 +80,89 @@ func reconcileResourceToNamespace(obj runtime.Object, controller string, nsListe
+		return fmt.Errorf("cannot create membership roles for unsupported type %T", v)
+	}
+
+	ownerRef, err := newOwnerReference(obj)
```
Does this work for `Projects`? `ClusterRoles` are non-namespaced and `Projects` are namespaced. See https://github.com/rancher/rancher/pull/48682#discussion_r1919286170

raulcabello commented on pkg/controllers/management/auth/project_cluster/common_test.go:
```diff
@@ -0,0 +1,174 @@
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			ctrl := gomock.NewController(t)
```
this could be outside of the for loop, so we avoid creating a new controller in each iteration.

JonCrowther replied:
Added in b88c01d892996f7f2b83622717cd927020e92387

raulcabello commented on pkg/controllers/management/auth/project_cluster/common_test.go:
```diff
@@ -0,0 +1,174 @@
+			crClient := fake.NewMockNonNamespacedControllerInterface[*rbacv1.ClusterRole, *rbacv1.ClusterRoleList](ctrl)
+
+			if tt.wantedCRs != nil {
+				crClient.EXPECT().Get(gomock.Any(), gomock.Any()).Return(nil, errNotFound).AnyTimes()
+				crClient.EXPECT().Create(gomock.Any()).DoAndReturn(func(c *rbacv1.ClusterRole) (*rbacv1.ClusterRole, error) {
+					assert.Equal(t, tt.wantedCRs[0], c)
+					return nil, nil
+				})
+				crClient.EXPECT().Create(gomock.Any()).DoAndReturn(func(c *rbacv1.ClusterRole) (*rbacv1.ClusterRole, error) {
+					assert.Equal(t, tt.wantedCRs[1], c)
+					return nil, nil
+				})
```
We should avoid gomock.Any() and AnyTimes() if possible to have more accurate tests. This way we also verify the parameters passed and the times the mock methods are called

```suggestion
				crClient.EXPECT().Get(tt.wantedCRs[0].Name, metav1.GetOptions{}).Return(nil, errNotFound)
				crClient.EXPECT().Get(tt.wantedCRs[1].Name, metav1.GetOptions{}).Return(nil, errNotFound)
				crClient.EXPECT().Create(tt.wantedCRs[0])
				crClient.EXPECT().Create(tt.wantedCRs[1])
```

JonCrowther replied:
Fixed in b88c01d892996f7f2b83622717cd927020e92387

raulcabello commented on pkg/controllers/management/auth/roletemplates/crtb_handler.go:
```diff
@@ -0,0 +1,278 @@
+package roletemplates
```
can you add tests for this file?

raulcabello commented on pkg/controllers/management/auth/roletemplates/prtb_handler_test.go:
```diff
@@ -0,0 +1,385 @@
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			ctrl := gomock.NewController(t)
```
can you add `t.Parallel()` so we run tests in parallel. This applies to other tests, same as moving  `ctrl := gomock.NewController(t)` outside of the for loop

JonCrowther replied:
Fixed in b88c01d892996f7f2b83622717cd927020e92387

raulcabello commented on pkg/controllers/managementuser/rbac/roletemplates/common.go:
```diff
@@ -0,0 +1,78 @@
+	return err
+}
+
+// deleteServiceAccountImpersonator checks if there are any CRBTs or PRTBs for this user. If there are none, remove their Service Account Impersonator.
```
nit:
```suggestion
// deleteServiceAccountImpersonator checks if there are any CRTBs or PRTBs for this user. If there are none, remove their Service Account Impersonator.
```

raulcabello reviewed (commented):
I run it locally, and I found a few things:

- User can't see cluster in the UI when permission to a project is given
- User can still see the cluster after the permissions are removed. I gave a user cluster-owner, then when removing the permission the cluster was still displayed in the UI (less menu items though)

aruiz14 commented on pkg/features/feature.go:
```diff
@@ -150,6 +150,12 @@ var (
 		true,
 		false,
 		true)
+	AggregatedRoleTemplates = newFeature(
+		"aggregated-roletemplates",
+		"[Experimental] Make RoleTemplates use aggregation for generated RBAC roles",
+		false,
+		false,
+		true)
```
I've looked into how to lock feature flags on install. You can find and possibly cherry-pick `23add5022e98a1bde323fba124c2d29c94f5d5b9` from my branch:
https://github.com/joncrowther/rancher/compare/aggregated-cluster-roles...aruiz14:rancher:aggregated-cluster-roles
![image](https://github.com/user-attachments/assets/e4bb8f73-7285-4256-85b1-74c3cb312e49)

JonCrowther replied:
Added in 638d47f382d9a098c3dc5983299c2a587e3fce9b

JonCrowther commented:
> I run it locally, and I found a few things:
> 
>     * User can't see cluster in the UI when permission to a project is given
> 
>     * User can still see the cluster after the permissions are removed. I gave a user cluster-owner, then when removing the permission the cluster was still displayed in the UI (less menu items though)

After some testing I discovered I wasn't adding or removing cluster and project membership correctly. I believe I have corrected it in the most recent changes. Feel free to test for yourself :)

alegrey91 reviewed (approved):
LGTM

raulcabello reviewed (changes requested):
It's looking good, but I found two issues that needs fixing:

- When granting a user `cluster-owner` permission to a downstream cluster, he also gets `cluster-owner` permission in the local cluster. The right CRB (ClusterRole/cluster-owner-cluster-mgmt-aggregator) is created, but for some reason this CRB is deleted and CRB pointing to ClusterRole/cluster-owner-aggregator is created in the local cluster.

- User can't see project in downstream cluster after he is granted project-owner permission. It looks like bindings to project owner or member in local cluster are not created

JonCrowther commented:
> It's looking good, but I found two issues that needs fixing:
> 
>     * When granting a user `cluster-owner` permission to a downstream cluster, he also gets `cluster-owner` permission in the local cluster. The right CRB (ClusterRole/cluster-owner-cluster-mgmt-aggregator) is created, but for some reason this CRB is deleted and CRB pointing to ClusterRole/cluster-owner-aggregator is created in the local cluster.
> 
>     * User can't see project in downstream cluster after he is granted project-owner permission. It looks like bindings to project owner or member in local cluster are not created

@raulcabello I believe I've fixed both of those issues.

The CRTB and PRTB `managementuser` controllers were always being run in the local cluster, so they would overwrite the `management` controllers. I've added a check to ensure the cluster the controllers are on matches the CRTB/PRTB cluster.

For PRTBs, the issue was that the role reference I was using was incorrect. It was referring to a Role Binding when the new reference is a Cluster Role Binding.

raulcabello commented:
> The CRTB and PRTB `managementuser` controllers were always being run in the local cluster, so they would overwrite the `management` controllers. I've added a check to ensure the cluster the controllers are on matches the CRTB/PRTB cluster.

Good catch! you are right controller were always being run in the local cluster (and other clusters). The check you added fixed the issue. Thanks!

> For PRTBs, the issue was that the role reference I was using was incorrect. It was referring to a Role Binding when the new reference is a Cluster Role Binding.

There seems to be an issue with PRTBs. A new user with project-owner can't see the project. I see this in the logs:

```
2025/02/04 11:35:58 [ERROR] error syncing 'p-g49n5/creator-project-owner': handler cluster-prtb-remove-handler: Operation cannot be fulfilled on projectroletemplatebindings.management.cattle.io "creator-project-owner": StorageError: invalid object, Code: 4, Key: /registry/management.cattle.io/projectroletemplatebindings/p-g49n5/creator-project-owner, ResourceVersion: 0, AdditionalErrorMsg: Precondition failed: UID in precondition: af4f83dc-4fe4-478d-9ea1-6a4e3c2e93c2, UID in object meta: , handler cluster-prtb-remove-handler: Operation cannot be fulfilled on projectroletemplatebindings.management.cattle.io "creator-project-owner": StorageError: invalid object, Code: 4, Key: /registry/management.cattle.io/projectroletemplatebindings/p-g49n5/creator-project-owner, ResourceVersion: 0, AdditionalErrorMsg: Precondition failed: UID in precondition: af4f83dc-4fe4-478d-9ea1-6a4e3c2e93c2, UID in object meta: , requeuing
2025/02/04 11:35:59 [ERROR] error syncing 'p-g49n5/prtb-dbdsp': handler cluster-prtb-remove-handler: Operation cannot be fulfilled on projectroletemplatebindings.management.cattle.io "prtb-dbdsp": StorageError: invalid object, Code: 4, Key: /registry/management.cattle.io/projectroletemplatebindings/p-g49n5/prtb-dbdsp, ResourceVersion: 0, AdditionalErrorMsg: Precondition failed: UID in precondition: a82ecaf9-a954-4103-9be1-5419a6129532, UID in object meta: , handler cluster-prtb-remove-handler: Operation cannot be fulfilled on projectroletemplatebindings.management.cattle.io "prtb-dbdsp": StorageError: invalid object, Code: 4, Key: /registry/management.cattle.io/projectroletemplatebindings/p-g49n5/prtb-dbdsp, ResourceVersion: 0, AdditionalErrorMsg: Precondition failed: UID in precondition: a82ecaf9-a954-4103-9be1-5419a6129532, UID in object meta: , requeuing
```

aruiz14 commented on pkg/rbac/common.go:
```diff
@@ -3,28 +3,51 @@ package rbac
+	promotedSuffix                    = "promoted"
+	clusterManagementPlaneSuffix      = "cluster-mgmt"
+	projectManagementPlaneSuffix      = "project-mgmt"
+	RBACApiGroup                      = "rbac.authorization.k8s.io"
```
This is already defined in [`k8s.io/api/rbac/v1`](https://pkg.go.dev/k8s.io/api/rbac/v1#pkg-constants) as `GroupName`

aruiz14 commented on pkg/controllers/managementuser/rbac/roletemplates/crtb_handler.go:
```diff
@@ -0,0 +1,204 @@
+	return returnError
+}
+
+var timeNow = func() time.Time {
+	return time.Now()
+}
```
Abstracting this behind a `Clock` interface would be cleaner. `k8s.io/utils` includes its own implementation that you can use: https://pkg.go.dev/k8s.io/utils/clock#Clock

JonCrowther replied:
I kept it the same as before. If we do change this to a `Clock` interface, I'd prefer to leave that to another PR
